tutorials/iot/graphical-snaps/graphical-snaps.md

```diff
@@ -0,0 +1,608 @@
+
+```
+gerry@ubuntu:/home/gerry/snaps/glmark2-wayland$ ls $SNAP/usr/share/glmark2
+models    shaders  textures
+```
+
+This is extremely common when snapping applications, therefore there are a few approaches to solving this:
+
+
+### Your application may read an environment variable
+Your application may read an environment variable that specifies where it should look for those resources. In that case, adjust your YAML file to add it like this:
+
+```yaml
+apps:
+  glmark2-wayland:
+  command: "usr/bin/glmark2-wayland"
+  environment:
+    RESOURCES_PATH: $SNAP/usr/share/glmark2
+```
+
+In our case, glmark2-wayland has the path “/usr/share/glmark2” hard-coded in, so this is not going to work for us.
+
+
+### Changing the application
+
+Sometimes you can edit/recompile the application to add the environment variable mentioned above. If it is your own code you are snapping, this is a good approach.
+
+We could do this, but glmark2-wayland is not our own code and this adds an unnecessary maintenance overhead. 
+
+### Using layouts for hard-coded paths
+
+This experimental snapd feature bind-mounts directories inside the snap into any location, so that the binaries’ hard-coded paths are correct. To use, add this to the YAML file:
+
+```yaml
+passthrough:
+  layout:
+    /usr/share/glmark2:
+      bind: $SNAP/usr/share/glmark2
+```
+
+## First Pass Snapping (resumed)
+duration: 3:00
+
+For this guide we are going to use “layouts” frequently whenever paths are hard-coded into binaries. So adding the snippet above, our YAML becomes
+
+```yaml
+name: glmark2-wayland
+version: 0.1
+summary: GLMark2 on Wayland
+description: |
+  GLMark2 on Wayland
+confinement: strict
+grade: devel
+
+apps:
+  glmark2-wayland:
+    command: "usr/bin/glmark2-wayland"
+    plugs:
+      - opengl
+      - wayland
+
+parts:
+  glmark2-wayland:
+    plugin: nil
+    stage-packages:
+      - glmark2-wayland
+
+passthrough:
+  layout:
+    /usr/share/glmark2:
+      bind: $SNAP/usr/share/glmark2
+```
+
+“snapcraft cleanbuild” this and install:
+
+```bash
+snapcraft cleanbuild
+sudo snap install --devmode ./glmark2-wayland_0.1_amd64.snap
+```
+This gets an error message:
+```
+error: cannot install snap file: cannot use experimental 'layouts' feature, set option
+       'experimental.layouts' to true and try again
+```
+This is snapd making sure you understand that layouts are an experimental feature. We have to tell snapd we know about that:
+
+```bash
+sudo snap set core experimental.layouts=true
+```
+
+and from now on, snapd will let us install snaps using layouts.
+
+positive
+: You may have noticed that snapcraft prints this scary message:
+The 'passthrough' property is being used to propagate experimental properties to snap.yaml that have not been validated. The snap cannot be released to the store.
+The policy on this has changed, these snaps can indeed be released to the store, but after manual verification.
```

Comment:
`snapcraft` from snap doesn't print that any more.

Reply:
Weird. I've installed the snap (and removed the deb) but I still see this message.

Reply:
Ah it's probably me using edge, then. There's no beta or candidate and suggesting edge is probably not a good idea. Maybe just add a comment that this message won't be there on snapcraft newer than `2.42.1`?

Reply:
Actually, the revised message isn't scary. And we now "snap install", so I'll drop this note.